src/plugins/es_ui_shared/public/request/use_request.ts

```diff
@@ -0,0 +1,162 @@
+  const pollIntervalIdRef = useRef<any>(null);
+
+  const clearPollInterval = useCallback(() => {
+    if (pollIntervalIdRef.current) {
+      clearTimeout(pollIntervalIdRef.current);
+      pollIntervalIdRef.current = null;
+    }
+  }, []);
+
+  // Convert our object to string to be able to compare them in our useMemo,
+  // allowing the consumer to freely passed new objects to the hook on each
+  // render without requiring them to be memoized.
+  const queryStringified = query ? JSON.stringify(query) : undefined;
+  const bodyStringified = body ? JSON.stringify(body) : undefined;
+
+  const requestBody = useMemo(() => {
+    return {
+      path,
+      method,
+      query: queryStringified ? query : undefined,
+      body: bodyStringified ? body : undefined,
+    };
+    // queryStringified and bodyStringified stand in for query and body as dependencies.
+    /* eslint-disable-next-line react-hooks/exhaustive-deps */
+  }, [path, method, queryStringified, bodyStringified]);
+
+  const sendRequest = useCallback(async () => {
+    // If we're on an interval, this allows us to reset it if the user has manually requested the
+    // data, to avoid doubled-up requests.
+    clearPollInterval();
+
+    // Pull off request ID and increment request count.
+    const requestId = requestCountRef.current++;
+
+    // We don't clear error or data, so it's up to the consumer to decide whether to display the
+    // "old" error/data or loading state when a new request is in-flight.
+    setIsLoading(true);
+
+    const response = await sendStatelessRequest<D, E>(httpClient, requestBody);
+    const { data: serializedResponseData, error: responseError } = response;
+
+    const isOutdatedRequest = requestId !== requestCountRef.current - 1;
+    const isUnmounted = isMounted.current === false;
+
+    // Ignore outdated or irrelevant data.
+    if (isOutdatedRequest || isUnmounted) {
+      return { data: null, error: null };
+    }
+
+    setError(responseError);
+    // If there's an error, keep the data from the last request in case it's still useful to the user.
+    if (!responseError) {
+      const responseData = deserializer
+        ? deserializer(serializedResponseData)
+        : serializedResponseData;
+      setData(responseData);
+    }
+    // Setting isLoading to false also acts as a signal for scheduling the next poll request.
+    setIsLoading(false);
+
+    return { data: serializedResponseData, error: responseError };
+  }, [requestBody, httpClient, deserializer, clearPollInterval]);
+
+  const scheduleRequest = useCallback(() => {
+    // If there's a scheduled poll request, this new one will supersede it.
+    clearPollInterval();
+
+    // Schedule next poll request.
+    if (pollIntervalMs) {
+      pollIntervalIdRef.current = setTimeout(sendRequest, pollIntervalMs);
+    }
+  }, [pollIntervalMs, sendRequest, clearPollInterval]);
+
+  // Send the request on component mount and whenever the dependencies of sendRequest() change.
```

Comment:
nit: same here about the unnecessary comment (here you are basically describing the `useEffect` lifecycle.

Reply:
I've removed the other comments you pointed out. I think this is the one comment I'm going to keep, just to really emphasize "the dependencies of sendRequest()" bit. It seems super obvious after staring at this code for hours, but I think the comment is more for someone who hasn't already spent a ton of time looking at the code. :)